File: claude/.claude/hooks/dangerous-command-guard.py
#!/usr/bin/env python3
"""PreToolUse hook (Bash): hard-block catastrophic / irreversible commands so a
reflexive permission-approval can't run them.

The problem: permission-prompt fatigue → commands approved without reading. This
moves the genuinely dangerous decision OFF the reflexive click: the command is
denied with a loud reason, and the user runs it deliberately via `!` if truly
intended. Best-effort pattern match, NOT a sandbox — it catches the classic
foot-guns, not every possible one.

FAIL-OPEN: any error → no output → command runs (subject to normal permissions).
"""
import sys
import json
import re

# (pattern, human label). Case-insensitive. Scoped to clearly destructive forms.
DANGER = [
    (r"\brm\s+(?:-\S+\s+)*-\S*r\S*f|\brm\s+(?:-\S+\s+)*-\S*f\S*r|\brm\s+-[rf]\s+-[rf]\b", "recursive force delete (rm -rf)"),
    (r"(?:^|\s|/)rm\s+-[a-z]*r[a-z]*\s+/(?:\s|$|\*)", "recursive delete from filesystem root"),
    (r"\bgit\s+push\b[^|;&\n]*(?:--force(?!-with-lease)\b|\s-f\b)", "git force-push"),
    (r"\bgit\s+reset\s+--hard\b", "git reset --hard (discards work)"),
    (r"\bgit\s+clean\s+-\S*f", "git clean -f (deletes untracked)"),
    (r"\b(?:migrate:fresh|migrate:reset|db:wipe)\b", "destructive DB migration (drops all tables)"),
    (r"\b(?:rails\s+db:drop|prisma\s+migrate\s+reset|sequelize\s+db:drop)\b", "destructive DB reset"),
    (r"\bDROP\s+(?:TABLE|DATABASE|SCHEMA)\b|\bTRUNCATE\s+(?:TABLE\s+)?\w", "destructive SQL (DROP/TRUNCATE)"),
    (r"\bdd\b[^|;&\n]*\bof=", "dd (raw disk/file overwrite)"),
    (r"\bmkfs\b|>\s*/dev/(?:sd|nvme|disk)|\bof=/dev/", "write to block device / format"),
    (r"\bfind\b[^|;&\n]*(?:-delete\b|-exec\s+rm\b)", "find with -delete / -exec rm"),
    (r"(?:curl|wget)\b[^|\n]*\|\s*(?:sudo\s+)?(?:sh|bash|zsh)\b", "pipe remote script into a shell"),
    (r"\bchmod\s+-R\s+0?777\b", "chmod -R 777"),
    (r"\bterraform\s+destroy\b", "terraform destroy"),
    (r"\bdocker\s+system\s+prune\b|\bdocker\s+volume\s+rm\b", "docker destructive prune / volume rm"),
    (r"(?:^|[\s;&|(])aws\s+\S", "the AWS CLI — disallowed for the agent (run it yourself if needed)"),
    (r":\(\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;\s*:", "fork bomb"),
]
DANGER = [(re.compile(p, re.I), label) for p, label in DANGER]


def emit_deny(reason):
    json.dump({"hookSpecificOutput": {
        "hookEventName": "PreToolUse",
        "permissionDecision": "deny",
        "permissionDecisionReason": reason,
    }}, sys.stdout)


# Values of human-message flags are prose, never a command — strip them before
# scanning so a commit message / PR body that merely MENTIONS a dangerous string
# ("… blocks rm -rf …") doesn't false-positive. Only -m/--message/--body; NOT -c
# or --body-file, so genuinely dangerous quoted payloads (psql -c 'DROP …') and
# real invocations stay scanned.
MSG_FLAG = re.compile(
    r"(?:-m|--message|--body)(?:=|\s+)(?:\"(?:[^\"\\]|\\.)*\"|'[^']*'|\S+)"
)


def strip_message_flags(cmd):
    return MSG_FLAG.sub(" ", cmd)


def main():
    data = json.load(sys.stdin)
    if data.get("tool_name") != "Bash":
        return
    cmd = (data.get("tool_input") or {}).get("command")
    if not isinstance(cmd, str) or not cmd.strip():
        return
    scan = strip_message_flags(cmd)
    for rx, label in DANGER:
        if rx.search(scan):
            emit_deny(
                f"Blocked by the command guard — {label}. Denied so it can't run on a "
                f"reflexive approval. If you genuinely intend it, run it yourself via the "
                f"`!` prefix, or restate it explicitly and I'll explain exactly what it does first."
            )
            return


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass  # fail-open: never block on a hook bug

File: clanker/config/hooks/dangerous-command-guard.ts
import type { HookResult, HookCtx, HookInput } from "./hook-utils"

// git/gh global options tolerated before the subcommand (git -C path, -c k=v, --long).
const GITX = String.raw`(?:-C\s+\S+\s+|-c\s+\S+\s+|--\S+\s+|-\w+\s+)*`

const DANGER: [RegExp, string][] = [
    [
        new RegExp(String.raw`\bgit\s+${GITX}push\b[^|;&\n]*(?:--force(?!-with-lease\b|-if-includes\b)|\s-f\b)`, "i"),
        "git force-push",
    ],
    [new RegExp(String.raw`\bgit\s+${GITX}reset\s+--hard\b`, "i"), "git reset --hard (discards work)"],
    [new RegExp(String.raw`\bgit\s+${GITX}clean\s+-\S*f`, "i"), "git clean -f (deletes untracked)"],
    [/\b(?:migrate:fresh|migrate:reset|db:wipe)\b/i, "destructive DB migration (drops all tables)"],
    [/\b(?:rails\s+db:drop|prisma\s+migrate\s+reset|sequelize\s+db:drop)\b/i, "destructive DB reset"],
    [/\bDROP\s+(?:TABLE|DATABASE|SCHEMA)\b|\bTRUNCATE\s+(?:TABLE\s+)?\w/i, "destructive SQL (DROP/TRUNCATE)"],
    [/\bdd\b[^|;&\n]*\bof=/i, "dd (raw disk/file overwrite)"],
    [/\bmkfs\b|>\s*\/dev\/(?:sd|nvme|disk)|\bof=\/dev\//i, "write to block device / format"],
    [/\bfind\b[^|;&\n]*(?:-delete\b|-exec\s+rm\b)/i, "find with -delete / -exec rm"],
    [/(?:curl|wget)\b[^|\n]*\|\s*(?:sudo\s+)?(?:sh|bash|zsh)\b/i, "pipe remote script into a shell"],
    [/\bterraform\s+destroy\b/i, "terraform destroy"],
    [/\bdocker\s+system\s+prune\b|\bdocker\s+volume\s+rm\b/i, "docker destructive prune / volume rm"],
    [/(?:^|[\s;&|(\\`])(?:\S*\/)?aws(?=\s|$)/i, "the AWS CLI — disallowed for the agent (run it yourself if needed)"],
    [/:\(\)\s*\{\s*:\s*\|\s*:\s*&\s*\}\s*;\s*:/, "fork bomb"],
]

const MSG_FLAG = /(?:-m|--message|--body)(?:=|\s+)(?:"(?:[^"\\]|\\.)*"|'[^']*'|\S+)/g
const SUBST = /\$\(([^)]*)\)|`([^`]*)`|\$\{([^}]*)\}/g

const stripMessageFlags = (cmd: string) => cmd.replace(MSG_FLAG, " ")

function substitutions(cmd: string): string {
    const out: string[] = []
    for (const m of cmd.matchAll(SUBST)) for (const g of [m[1], m[2], m[3]]) if (g) out.push(g)
    return out.join(" ")
}

const base = (t: string) => t.split("/").pop() ?? t

// An `rm` whose flags include BOTH recursive and force, any order (not `git rm`).
function rmRecursiveForce(text: string): boolean {
    const toks = text.split(/\s+/).filter(Boolean)
    for (let i = 0; i < toks.length; i++) {
        if (base(toks[i]) !== "rm") continue
        if (i > 0 && base(toks[i - 1]) === "git") continue
        let short = ""
        const longs: string[] = []
        for (const t of toks.slice(i + 1)) {
            if (t.startsWith("--")) longs.push(t)
            else if (t.startsWith("-") && t.length > 1) short += t.slice(1)
            else break
        }
        const hasR = short.toLowerCase().includes("r") || longs.includes("--recursive")
        const hasF = short.toLowerCase().includes("f") || longs.includes("--force")
        if (hasR && hasF) return true
    }
    return false
}

function chmod777Recursive(text: string): boolean {
    const toks = text.split(/\s+/).filter(Boolean)
    for (let i = 0; i < toks.length; i++) {
        if (base(toks[i]) !== "chmod") continue
        const rest = toks.slice(i + 1)
        const recursive = rest.some(
            (x) => x === "--recursive" || (x.startsWith("-") && !x.startsWith("--") && x.includes("R")),
        )
        const mode = rest.some((x) => x === "777" || x === "0777" || x === "a+rwx")
        if (recursive && mode) return true
    }
    return false
}

function scan(text: string): string | null {
    if (rmRecursiveForce(text)) return "recursive force delete (rm -rf)"
    if (chmod777Recursive(text)) return "chmod -R 777"
    for (const [rx, label] of DANGER) if (rx.test(text)) return label
    return null
}

export async function run(input: HookInput, _ctx: HookCtx): Promise<HookResult> {
    const cmd = input.command
    if (typeof cmd !== "string" || !cmd.trim()) return { kind: "none" }
    for (const text of [stripMessageFlags(cmd), substitutions(cmd)]) {
        if (!text.trim()) continue
        const label = scan(text)
        if (label) {
            return {
                kind: "deny",
                reason: `Blocked by the command guard — ${label}. Denied so it can't run on a reflexive approval. If you genuinely intend it, run it yourself via the \`!\` prefix, or restate it explicitly and I'll explain exactly what it does first.`,
            }
        }
    }
    return { kind: "none" }
}
